src/server/master_server/master_metadata_service_impl.cc

```diff
@@ -12,13 +11,48 @@ using protos::grpc::OpenFileRequest;
 namespace gfs {
 namespace service {
 
+inline server::MetadataManager* MasterMetadataServiceImpl::metadata_manager() {
+  return server::MetadataManager::GetInstance();
+}
+
+grpc::Status MasterMetadataServiceImpl::HandleFileCreation(
+    const protos::grpc::OpenFileRequest* request,
+    protos::grpc::OpenFileReply* reply) {
+  const std::string& filename(request->filename());
+  google::protobuf::util::Status status(
+      metadata_manager()->CreateFileMetadata(filename)); 
+  return common::utils::ConvertProtobufStatusToGrpcStatus(status);
+}
+
+grpc::Status MasterMetadataServiceImpl::HandleFileChunkRead(
+    const protos::grpc::OpenFileRequest* request,
+    protos::grpc::OpenFileReply* reply) {
+  return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "needs implementation");
+}
+
+grpc::Status MasterMetadataServiceImpl::HandleFileChunkWrite(
+    const protos::grpc::OpenFileRequest* request,
+    protos::grpc::OpenFileReply* reply) {
+  return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "needs implementation");
+}
+
 grpc::Status MasterMetadataServiceImpl::OpenFile(ServerContext* context,
                                                  const OpenFileRequest* request,
                                                  OpenFileReply* reply) {
-  // TODO(everyone): implement the GFS master server logic here
-  *reply->mutable_request() = *request;
-  reply->set_chunk_version(1);
-  return grpc::Status::OK;
+  // Dispatch different mode to different handle function
+  switch (request->mode()) {
+    case OpenFileRequest::CREATE:
+      return HandleFileCreation(request, reply);
+    case OpenFileRequest::READ:
+      return HandleFileChunkRead(request, reply);
+    case OpenFileRequest::WRITE:
+      return HandleFileChunkWrite(request, reply);
+    default:
+      return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "invalid mode");
+  }
+  
+  // Should not come to this line as all cases have been covered above 
+  return grpc::Status(grpc::StatusCode::UNKNOWN, "internal error");
```

Comment:
is it possible to not have this line at all? or compiler will error?

Reply:
Traditional compiler may throw warning when not seeing a return at the end, but I have tried it without this line and it seems alright. So I removed it.

Reply:
yeah. i think compiler nowadays can do program analysis on the switch statement when there's a default block